Loading TI Dashboard in Prod Times Out

We're getting the following error on Seattle prod when a TI user logs in and gets redirected to the TI Dashboard `/admin/tiDash?page=1`.

Looks related to the ProgramModel.

It's also slow when a CiviForm Admin tries to load `/admin/tiGroups` may or may not be related.

The rest of the application seems to be working fine.


```
play.api.UnexpectedException: Unexpected exception[RuntimeException: javax.persistence.PersistenceException: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms.] 
    at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:358) 
    at play.api.http.HttpErrorHandlerExceptions.throwableToUsefulException(HttpErrorHandler.scala) 
    at play.http.DefaultHttpErrorHandler.throwableToUsefulException(DefaultHttpErrorHandler.java:244) 
    at play.http.DefaultHttpErrorHandler.onServerError(DefaultHttpErrorHandler.java:180) 
    at controllers.ErrorHandler.onServerError(ErrorHandler.java:106) 
    at play.core.j.JavaHttpErrorHandlerAdapter.onServerError(JavaHttpErrorHandlerAdapter.scala:32) 
    at play.core.server.AkkaHttpServer$$anonfun$2.applyOrElse(AkkaHttpServer.scala:441) 
    at play.core.server.AkkaHttpServer$$anonfun$2.applyOrElse(AkkaHttpServer.scala:433) 
    at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:490) 
    at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63) 
    at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100) 
    at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18) 
    at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94) 
    at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100) 
    at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49) 
    at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48) 
    at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source) 
    at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source) 
    at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source) 
    at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source) 

Caused by: java.lang.RuntimeException: javax.persistence.PersistenceException: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms. 
    at org.pac4j.play.java.SecureAction.call(SecureAction.java:79) 
    at play.core.j.JavaAction.$anonfun$apply$8(JavaAction.scala:175) 
    at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:687) 
    at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467) 
    at play.core.j.HttpExecutionContext.$anonfun$execute$1(HttpExecutionContext.scala:64) 
    at play.api.libs.streams.Execution$trampoline$.executeScheduled(Execution.scala:109) 
    at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:71) 
    at scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:429) 
    at scala.concurrent.impl.Promise$DefaultPromise.submitWithValue(Promise.scala:338) 
    at scala.concurrent.impl.Promise$DefaultPromise.dispatchOrAddCallbacks(Promise.scala:312) 
    at scala.concurrent.impl.Promise$DefaultPromise.flatMap(Promise.scala:176) 
    at play.api.libs.streams.Accumulator$.$anonfun$futureToSink$1(Accumulator.scala:231) 
    at akka.stream.impl.Transform.apply(TraversalBuilder.scala:159) 
    at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:522) 
    at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:448) 
    at akka.stream.impl.PhasedFusingActorMaterializer.materialize(PhasedFusingActorMaterializer.scala:440) 
    at akka.stream.scaladsl.RunnableGraph.run(Flow.scala:764) 
    at akka.stream.scaladsl.Source.runWith(Source.scala:118) 
    at akka.stream.javadsl.Source.runWith(Source.scala:1075) 
    at play.libs.streams.Accumulator$FlattenSubscriber.lambda$handler$1(Accumulator.java:285) 

Caused by: javax.persistence.PersistenceException: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms. 
    at io.ebeaninternal.server.transaction.TransactionFactoryBasic.createReadOnlyTransaction(TransactionFactoryBasic.java:34) 
    at io.ebeaninternal.server.transaction.TransactionManager.createReadOnlyTransaction(TransactionManager.java:337) 
    at io.ebeaninternal.server.core.DefaultServer.createReadOnlyTransaction(DefaultServer.java:2177) 
    at io.ebeaninternal.server.core.OrmQueryRequest.initTransIfRequired(OrmQueryRequest.java:221) 
    at io.ebeaninternal.server.core.DefaultServer.findList(DefaultServer.java:1462) 
    at io.ebeaninternal.server.core.DefaultServer.findList(DefaultServer.java:1442) 
    at io.ebeaninternal.server.core.DefaultBeanLoader.executeQuery(DefaultBeanLoader.java:154) 
    at io.ebeaninternal.server.core.DefaultBeanLoader.loadBean(DefaultBeanLoader.java:137) 
    at io.ebeaninternal.server.core.DefaultServer.loadBean(DefaultServer.java:544) 
    at io.ebeaninternal.server.loadcontext.DLoadBeanContext$LoadBuffer.loadBean(DLoadBeanContext.java:207) 
    at io.ebean.bean.EntityBeanIntercept.loadBeanInternal(EntityBeanIntercept.java:875) 
    at io.ebean.bean.EntityBeanIntercept.loadBean(EntityBeanIntercept.java:852) 
    at io.ebean.bean.EntityBeanIntercept.preGetter(EntityBeanIntercept.java:946) 
    at models.ProgramModel._ebean_get_programDefinition(ProgramModel.java:6) 
    at models.ProgramModel._ebean_set_programDefinition(ProgramModel.java:1) 
    at models.ProgramModel.loadProgramDefinition(ProgramModel.java:267) 
    at jdk.internal.reflect.GeneratedMethodAccessor32.invoke(Unknown Source) 
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) 
    at java.base/java.lang.reflect.Method.invoke(Unknown Source) 
    at io.ebeaninternal.server.deploy.BeanLifecycleAdapterFactory$PostLoadAdapter.invoke(BeanLifecycleAdapterFactory.java:279) 

Caused by: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms. 
    at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:695) 
    at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:197) 
    at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:162) 
    at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100) 
    at play.db.ebean.DefaultEbeanConfig$EbeanConfigParser$WrappingDatasource.getConnection(DefaultEbeanConfig.java:158) 
    at io.ebeaninternal.server.transaction.TransactionFactoryBasic.createReadOnlyTransaction(TransactionFactoryBasic.java:27) 
    at io.ebeaninternal.server.transaction.TransactionManager.createReadOnlyTransaction(TransactionManager.java:337) 
    at io.ebeaninternal.server.core.DefaultServer.createReadOnlyTransaction(DefaultServer.java:2177) 
    at io.ebeaninternal.server.core.OrmQueryRequest.initTransIfRequired(OrmQueryRequest.java:221) 
    at io.ebeaninternal.server.core.DefaultServer.findList(DefaultServer.java:1462) 
    at io.ebeaninternal.server.core.DefaultServer.findList(DefaultServer.java:1442) 
    at io.ebeaninternal.server.core.DefaultBeanLoader.executeQuery(DefaultBeanLoader.java:154) 
    at io.ebeaninternal.server.core.DefaultBeanLoader.loadBean(DefaultBeanLoader.java:137) 
    at io.ebeaninternal.server.core.DefaultServer.loadBean(DefaultServer.java:544) 
    at io.ebeaninternal.server.loadcontext.DLoadBeanContext$LoadBuffer.loadBean(DLoadBeanContext.java:207) 
    at io.ebean.bean.EntityBeanIntercept.loadBeanInternal(EntityBeanIntercept.java:875) 
    at io.ebean.bean.EntityBeanIntercept.loadBean(EntityBeanIntercept.java:852) 
    at io.ebean.bean.EntityBeanIntercept.preGetter(EntityBeanIntercept.java:946) 
    at models.ProgramModel._ebean_get_programDefinition(ProgramModel.java:6) 
    at models.ProgramModel._ebean_set_programDefinition(ProgramModel.java:1) 


```

--- a/server/app/views/applicant/TrustedIntermediaryDashboardView.java
+++ b/server/app/views/applicant/TrustedIntermediaryDashboardView.java
@@ -41,11 +41,11 @@ import models.AccountModel;
 import models.ApplicantModel;
 import models.ApplicationModel;
 import models.TrustedIntermediaryGroupModel;
+import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import play.i18n.Messages;
 import play.mvc.Http;
 import play.twirl.api.Content;
-import repository.ProgramRepository;
 import repository.SearchParameters;
 import services.DateConverter;
 import services.MessageKey;
@@ -53,6 +53,8 @@ import services.PhoneValidationResult;
 import services.PhoneValidationUtils;
 import services.applicant.ApplicantData;
 import services.applicant.ApplicantPersonalInfo;
+import services.program.ProgramNotFoundException;
+import services.program.ProgramService;
 import services.ti.TrustedIntermediaryService;
 import views.BaseHtmlView;
 import views.HtmlBundle;
@@ -68,17 +70,19 @@ import views.style.StyleUtils;
 
 /** Renders a page for a trusted intermediary to manage their clients. */
 public class TrustedIntermediaryDashboardView extends BaseHtmlView {
+  private static final Logger logger =
+      LoggerFactory.getLogger(TrustedIntermediaryDashboardView.class);
   private final ApplicantLayout layout;
   private final DateConverter dateConverter;
-  private final ProgramRepository programRepository;
+  private final ProgramService programService;
   private static final PhoneNumberUtil PHONE_NUMBER_UTIL = PhoneNumberUtil.getInstance();
 
   @Inject
   public TrustedIntermediaryDashboardView(
-      ApplicantLayout layout, DateConverter dateConverter, ProgramRepository programRepository) {
+      ApplicantLayout layout, DateConverter dateConverter, ProgramService programService) {
     this.layout = checkNotNull(layout);
     this.dateConverter = checkNotNull(dateConverter);
-    this.programRepository = checkNotNull(programRepository);
+    this.programService = programService;
   }
 
   public Content render(
@@ -369,11 +373,24 @@ public class TrustedIntermediaryDashboardView extends BaseHtmlView {
     String programs =
         newestApplicantApplications.stream()
             .map(
-                application ->
-                    programRepository
-                        .getShallowProgramDefinition(application.getProgram())
+                application -> {
+                  try {
+                    return programService
+                        .getFullProgramDefinition(application.getProgram().id)
                         .localizedName()
-                        .getDefault())
+                        .getDefault();
+                  } catch (ProgramNotFoundException e) {
+                    // Since this is just trying to get a csv representation of the programs
+                    // put a placeholder string if this exception occurs. Realistically at
+                    // this area of CiviForm it "shouldn't" occur, but why leave it up to
+                    // fate to raise an exception.
+                    logger.error(
+                        "Unable to build complete string of programs. At least one program was not"
+                            + " found",
+                        e);
+                    return "<unknown>";
+                  }
+                })
             .collect(Collectors.joining(", "));
 
     return div(

